[IMP] website, website_sale, web_editor: add `s_floating_blocks` snippet

stefanorigano opened the pull request:
This PR introduce a new static snippet tailored for eCommerce categories and promos.

It also make the "Products" snippets category visible by default, allowing the injection of snippets that don't require `website_sale` to be installed.

task-4373543
part-of-task-4252024

Related:
- https://github.com/odoo/design-themes/pull/1054

| Screenshoot |
|--------|
| <img width="1340" alt="image" src="https://github.com/user-attachments/assets/f037f3df-5528-4372-9108-6b7d3eb5b49e"> | 



---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/189299.png)](https://mergebot.odoo.com/odoo/odoo/pull/189299)

chevalierv reviewed (commented):
Hello @stefanorigano 

Thanks for suggesting this new snippet! After discussing with @Jean-boje, I think that those changes should be added directly to `website`, as there is no specific feature to eCommerce.

Could you move your changes into the website module?

Have a nice day!

stefanorigano commented:
Hello @sobo-odoo 👋🙂
Could you have a look at this? Already approved by BOJE/LEBL. 

Note: we'll enable the "Products" category by default, even if `website_sale` it's not installed (similarly to newsletter snippets).

Thanks!

sobo-odoo commented:
@stefanorigano Wow this snippet is really cool! :scream::star_struck: I don't think I can review it unfortunately, as I am working on the refactoring and given that the PR is huge :grimacing:  Maybe ask BVR or QSM (or someone else) on Monday instead :slightly_smiling_face:

stefanorigano commented:
Thanks for the fast feedback @sobo-odoo 👍 
Ok, @bvr-odoo or @qsm-odoo  then.

Thanks in advance!

stefanorigano commented on addons/website/static/src/snippets/s_floating_blocks/000.js:
```diff
@@ -0,0 +1,172 @@
+import {throttleForAnimation} from "@web/core/utils/timing";
+import publicWidget from "@web/legacy/js/public/public_widget";
+
+const FloatingBlocks = publicWidget.Widget.extend({
```
Hello @qsm-odoo , no worries.

I've adapted the code to the new system and renamed the files according to the new convention. It seems to work well, except for a traceback when` _refreshPublicWidgets()` is called by the snippet's options (to replicate, simply add a new card).

![image](https://github.com/user-attachments/assets/6803cec9-8c80-4435-95a0-7f01fd0b09fe)

The error message is:

```console
Error: this.addListener can only be called after the interaction is started. Maybe move the call in the start method.
```
The listeners seem to be defined in the correct location tho, and work correctly until `_refreshPublicWidgets()` is called.

I'm not sure if this issue is due to my changes or if it reveals a general bug. I don't see any significant differences compared to other snippets using the same method, but I might be missing something.

Could you have a look? Thanks

stefanorigano commented:
Hello @qsm-odoo , I've updated the code. 

About my issue with `this._refreshPublicWidgets();` (sorry, I must have deleted my message by mistake), I've ultimately solved it  [here](https://github.com/odoo/odoo/pull/189299/files#diff-9691e7243f8efab4b4b0512510f6c54786ca7fceca9579b07325fd5af3e14e05R85) using `this.trigger_up('content_changed')`.

Let me know if you're up for review or if we should delegate.

Thanks

qsm-odoo commented on addons/website/static/src/snippets/s_floating_blocks/options.js:
```diff
@@ -0,0 +1,143 @@
+    /**
+     * @override
+     */
+    cleanUI() {
+        this.$target[0].style.transform = "";
+        this.$target[0].classList.remove("opacity-0", "transition-base");
+    },
```
This should not be necessary. If it is, it would need comments.

qsm-odoo commented on addons/website/static/src/snippets/s_floating_blocks/options.js:
```diff
@@ -0,0 +1,143 @@
+     * Injects a new Card.
+     */
+    addCard() {
+        const newCard = renderToElement("website.s_floating_blocks.new_card");
```
Not a fan of having the XML of those "card" on both the static side and python-XML side. In general, don't we rather use the last existing card and clone it? (like for carousels?) What do you think?

stefanorigano replied:
Yes, I see what you mean but I don't think that cloning is always good.
It works well for components like tabs and accordions, but it doesn’t fit all the other use cases.

-  preventing the removal of the last entry just to maintain a "cloneable" structure can be confusing for newcomers, especially when they have already removed multiple entries with the very same bottom and they can see only one entry at the time.
- can be misleading; they clicked "Add New," not "Clone" at the end 
- For "complex" entries (like these "cards"), users need a "fresh start" option as well IMO, since stripping down an entry manually using the editor can be tedious and error-prone. At the end when you create a new page you start from scratch (or use a template), you don't clone the last page... and sometimes snippets are more complex than asimple page.

As said cloning makes sense for accordion and tabs-like component tho, but IMO just because:
- all entries are visible on screen, so it's easy to understand what's going on.
- these entries are text-based and easy to edit

All considered I think that the approach used in this PR is ok, especially in term of UX.
Technically is probably simpler, 'cause by cloning we have to prevent the removal of the last entry, toggle the remove button `disabled` state, strip away inline styles etc... I mean, not a big deal but still.

qsm-odoo replied:
My comment was not about the UX, except maybe speaking of UX consistency. I agree with what you said, except I would actually go further: cloning when clicking on "+" / "Add item" / ... is always bad IMO (even for accordion, tabs, etc). It should only clone when... clicking the clone button.

But what I said was from a technical point of view: we clone because it allows to create a correct structure easily, without duplicating code. And since we do that some places, we really should do it everywhere by consistency.

But let's say it will be a question for later: we can probably review all "Add item" buttons to not clone anymore, and we can do that without having to duplicate views on both JS and python side (we can just preload the python ones to have them available in JS for instance), cc @lebl-odoo (I won't create a task about this, I let you judge if you agree we should just stop cloning stuff when the user did not ask to clone anything).

lebl-odoo replied:
It's not the first time we have discussion about this specific topic, I'll note the feedback for later. I'll have a look after the refactoring. Thanks for point it out @qsm-odoo .

qsm-odoo replied:
> Considering that we already generate content trough JS templates and that "clone is bad" , would you consider problematic to proceed with the current implementation and eventually adapt it to the new system when it will be implemented?

It's already what I said in my previous message indeed, I'll review asap

qsm-odoo commented on addons/website/static/src/snippets/s_floating_blocks/options.xml:
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates xml:space="preserve">
+    <t t-name="website.s_floating_blocks.alert.empty">
```
Don't indent the first level of XML files in website (add blank line around each template though)

qsm-odoo commented on addons/website/views/snippets/s_floating_blocks.xml:
```diff
@@ -0,0 +1,174 @@
+    <section class="s_floating_blocks pt32 pb32 o_cc o_cc2">
+        <div class="container">
+            <div class="d-flex flex-column gap-5">
+                <span class="s_floating_blocks_wrapper s_floating_blocks_wrapper_shadow d-contents rounded-3">
+
+                    <section data-name="Card"
```
`span > section` is not valid HTML

qsm-odoo commented on addons/website/views/snippets/s_floating_blocks.xml:
```diff
@@ -0,0 +1,174 @@
+
+                    <section data-name="Card"
+                        class="o_block s_col_no_resize position-sticky d-flex overflow-hidden py-5 o_cc o_cc5 oe_img_bg o_bg_img_center o_background_video"
+                        data-bg-video-src="https://player.vimeo.com/video/329369394?autoplay=1&amp;muted=1&amp;autopause=0&amp;controls=0&amp;loop=1"
+                        style="background-image:url(https://i.vimeocdn.com/video/774173730-b7a179626fb9429551c395a423571c56515633a3cdbf18261054f281a022b19d-d?mw=80&amp;q=85)">
```
Do you absolutely need to use a video by default? It will be annoying as we cannot add external references in tours.

stefanorigano replied:
I've replaced the "placeholder image" (prior to video starts) with a better static one.
I'd keep the video tho, it's nice to showcase that we provide these features :)

qsm-odoo commented on addons/website/views/snippets/s_floating_blocks.xml:
```diff
@@ -0,0 +1,174 @@
+
+                            <!-- Placeholder to ensure a minimum grid height -->
+                            <div class="o_grid_item oe_unremovable d-none g-col-lg-1 g-height-1 col-lg-1" aria-hidden="true" style="z-index: -1; grid-area: 11 / 12 / 12 / 13;">&#42;</div>
+                            <!-- // -->
```
Remove this kind of thing

qsm-odoo reviewed (changes requested):
A first quick look without testing, I'll test and check in details afterwards :pray: Also, you can squash all your commits.

stefanorigano commented on addons/web_editor/static/src/scss/wysiwyg_snippets.scss:
```diff
@@ -2912,6 +2912,21 @@ we-select.o_grid we-toggler {
+            &:hover [data-snippet="s_floating_blocks"] .o_block:nth-child(1) {
+                margin: 3% 0 -30%;
+                transform: scale(.95);
+            }
```
Yes, it was already the case indeed.
"Fixed" in 7c8130e3427aa1d99edb8865d799d26d3ada3df1, I've added proper comments as well.

stefanorigano commented on addons/website/static/src/snippets/s_floating_blocks/options.js:
```diff
@@ -0,0 +1,143 @@
+    start() {
+        this.wrapperEl = this.$target[0].querySelector(".s_floating_blocks_wrapper");
+        this.alertEl = this.wrapperEl.querySelector(".s_floating_blocks_alert_empty");
+        this.boxes = this.wrapperEl.querySelectorAll(".o_block");
```
Yes, you can.
When a block is cloned, `FloatingBlocksBlock` triggers an `option_update` that call  `_validateBoxesNumber` trough `notify`.

stefanorigano commented:
Hello @qsm-odoo , thanks for your review.
I've updated the code accordingly, "resolved" the most obvious comments and replied where necessary.

In therms of git history, at the current state we have 3 commits:
- Enable the dialog's  "Products" tab for static snippets
- Add the new snippet
- Add a glob pattern for snippet' scss in edit mode (for the modal preview).

Thanks.

qsm-odoo commented:
Trying out the snippet with looking at the code now, it looks cool.

I have this quirky overlay placement:
![image](https://github.com/user-attachments/assets/a99c63c4-918f-4c52-8bbe-7d6ef515d5b7)

It can happen in two cases:
- Clicking on a card, then simply scrolling a bit. The clicked card is "still there", then the overlay still hovers it. Maybe we could fix that on the overlay code itself (trying to check if the element is not "behind another" but not sure it would be easy/reliable (you can be behind and be visible)). But maybe there is a way to avoid that issue on snippet level @stefanorigano ? Maybe forcing hidden card to be actually display: none (in which case, the overlay would disappear)? That would actually make the transparent case a bit better, making fully covered cards disappear :shrug: 
![image](https://github.com/user-attachments/assets/60227926-8995-4ee9-9a13-c1bd522ed4d0)

- Adding a new card: when you click the button, we scroll to the new card... but we keep the old one enabled. Should not we either not scroll to the new item or enable the new item (I don't think enabling the new item is consistent with our other behaviors... but that's the one I would choose: it's weird to see option of cards on the right but those are about the old cards, not the new one). We could also scroll to the new item but disabling the old overlay :shrug: 

cc @lebl-odoo

stefanorigano commented:
> it looks cool.

@qsm-odoo , thanks and thanks  for your review.

> I have this quirky overlay placement:

Ooo thanks for noticing it... during the last iterations I've oversimplified the code and removed by mistake one part making cards to stack with a slight gap.  Edition was confusing in the scenario you described indeed.

It has been fixed and, when it's the case, it's now clear that the card selected is NOT the one on top.

| desktop | mobile |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/ed4f20c7-8386-481c-b882-a38ef7acdc8c) | ![image](https://github.com/user-attachments/assets/fdeb9114-c154-48f4-83dc-3902e31068f7) | 

---

> Adding a new card: when you click the button, we scroll to the new card... but we keep the old one enabled.[...]

Right, thanks. Solved by activating the newly created card.

---

// Other chnages

- Make use of CSS `contain` and `backface-visibility` to prove performance
- Decreased `dynamicContent` workload. `opacity` is now set during initialization, `top` when the header changes only.
- Use cached values to check if _isInViewport rather than using getBoundingClientRect
- Pre-compute transform values
- Fixed some missing some `El` and `Els` suffixes on variables name.

---

The diff -> https://github.com/odoo/odoo/compare/f8cea4138d45a260bf43c63292706af12fcd575d..8f3786a29ff820d076b619c83acb20f60db7f71b

---

Thanks!

qsm-odoo reviewed (changes requested):
@stefanorigano

- I push-forced with a rebase + squashing of your last 2 commits (since they cannot exist without each other, it should be only one commit)
- Then I pushed a "tmp" commit, meant to be squashed in the second one. It contains multiple changes:
  - Some linter issues + forcing comments on 80 chars per line (which is our convention)
  - Some stylistic changes (blank lines where I went them or not, variable names, ...
  - Override of function just calling super: useless
  - Your last commit contained CSS to change the grid-area in edit mode... I did not understand it: with it it is broken, without it, it seems to act fine?
  - Use of the `.noUpdate` in the interaction: as stated in the doc, the point is to avoid using it. If I understood what you wanted correctly: you needed it only because you wanted to check the snippet is in the viewport before performing computation. I'll often think it is useless, especially in this case where checking you are in the viewport might be as costly as performing the update TBH. It's also more code. So I simplified, not checking the viewport anymore. I think if we do need it, we'd rather have a `.inViewport` modifier or something like that at interaction engine-level (= why for this one and not for parallax for instance?).
  - The "_rebuildComponent" stuff: useless. The interaction are automatically updated after each option update, and if not (the onRemove case), you can call `_refreshPublicWidgets` just like before.
  - The "opacity" at the start: it was cleaned in the options. It's not its job, it's an interaction thing: the cleanup should be in the interaction. In this case, moving it in dynamicContent instead of the start does the trick.
  - Actually added a comment on that opacity: do we really need it? It seems fine, but why on this snippet and not the others?
  - Update of the `top`, same as opacity and transform: moved in the dynamicContent stuff. This allows for proper cleanup before save, not at option level.
  - In the new system, matching the OWL/backend, we don't use `_` function / `@private` anymore
  - "If top = previousTop, do nothing" -> removing this kind of stuff. It only makes the code harder to read for no real perf improvement
  - Maybe a few other minor things

Could you have a look / double-check I did not break anything? I'll then squash my changes and merge. Thanks :pray:

stefanorigano commented:
@qsm-odoo , 

My last iterations were focused on trying to simplifying `dynamicContent`,  as much as possible (`.noUpdate` , `inViewport`...) . Instinctively I thought that it was the right thing to do considering the amount of times onScroll was fired. Still, apparently I've worried too much 🙂
I've test both implementation performances and your version is not just simpler,  but also slightly faster indeed, thanks.
I still need to get used to this new system I think.

<details><summary>The test script</summary>
<p>

```js

(function() {
    // Performance monitoring variables
    const metrics = {
        scrollEvents: 0,
        scrollTimes: [],
        frames: 0,
        frameTimes: [],
        startTime: null,
        testRunning: false,
        testDuration: 10000, // 10 seconds by default
        scrollPattern: 'slow' // 'slow', 'fast', or 'erratic'
    };
    
    // Setup FPS monitoring
    let lastFrameTime = 0;
    function measureFrameRate() {
        if (!metrics.testRunning) return;
        
        const now = performance.now();
        if (lastFrameTime) {
            const frameTime = now - lastFrameTime;
            metrics.frameTimes.push(frameTime);
        }
        
        lastFrameTime = now;
        metrics.frames++;
        
        requestAnimationFrame(measureFrameRate);
    }
    
    // Intercept scroll events to measure performance
    const originalScroll = window.scroll;
    const originalScrollTo = window.scrollTo;
    
    function interceptScrollMethods() {
        window.scroll = function() {
            const start = performance.now();
            const result = originalScroll.apply(this, arguments);
            if (metrics.testRunning) {
                metrics.scrollEvents++;
                metrics.scrollTimes.push(performance.now() - start);
            }
            return result;
        };
        
        window.scrollTo = function() {
            const start = performance.now();
            const result = originalScrollTo.apply(this, arguments);
            if (metrics.testRunning) {
                metrics.scrollEvents++;
                metrics.scrollTimes.push(performance.now() - start);
            }
            return result;
        };
    }
    
    function restoreScrollMethods() {
        window.scroll = originalScroll;
        window.scrollTo = originalScrollTo;
    }
    
    // Scroll test patterns
    function runScrollPattern() {
        if (!metrics.testRunning) return;
        
        const elapsed = performance.now() - metrics.startTime;
        const progress = elapsed / metrics.testDuration;
        const documentHeight = Math.max(
            document.body.scrollHeight, 
            document.documentElement.scrollHeight
        );
        
        switch (metrics.scrollPattern) {
            case 'slow':
                // Smooth scroll from top to bottom
                window.scrollTo(0, progress * documentHeight);
                break;
                
            case 'fast':
                // Oscillating scroll (up and down quickly)
                window.scrollTo(0, documentHeight * Math.abs(Math.sin(progress * Math.PI * 8)));
                break;
                
            case 'erratic':
                // Random jumps
                if (Math.random() > 0.7) { // Only jump 30% of the time
                    window.scrollTo(0, Math.random() * documentHeight);
                }
                break;
        }
        
        if (elapsed < metrics.testDuration) {
            requestAnimationFrame(runScrollPattern);
        } else {
            endTest();
        }
    }
    
    // Start performance test
    function startTest(pattern = 'slow', duration = 10000) {
        console.log(`Starting ${pattern} scroll performance test for ${duration/1000} seconds...`);
        
        // Reset metrics
        metrics.scrollEvents = 0;
        metrics.scrollTimes = [];
        metrics.frames = 0;
        metrics.frameTimes = [];
        metrics.startTime = performance.now();
        metrics.testRunning = true;
        metrics.testDuration = duration;
        metrics.scrollPattern = pattern;
        
        // Reset scroll position
        window.scrollTo(0, 0);
        
        // Setup monitoring
        interceptScrollMethods();
        lastFrameTime = 0;
        
        // Start measurements
        measureFrameRate();
        runScrollPattern();
        
        // Create visual indicator
        const indicator = document.createElement('div');
        indicator.id = 'performance-test-indicator';
        indicator.style.cssText = 'position:fixed;bottom:20px;right:20px;background:rgba(0,0,0,0.8);color:white;padding:10px;z-index:9999;border-radius:4px;';
        indicator.textContent = `Testing: ${pattern} scroll (${duration/1000}s)`;
        document.body.appendChild(indicator);
        
        return `Test started: ${pattern} scroll for ${duration/1000} seconds`;
    }
    
    // End test and report results
    function endTest() {
        metrics.testRunning = false;
        restoreScrollMethods();
        
        // Calculate results
        const avgScrollTime = metrics.scrollTimes.length ? 
            metrics.scrollTimes.reduce((a, b) => a + b, 0) / metrics.scrollTimes.length : 0;
        
        const avgFrameTime = metrics.frameTimes.length ? 
            metrics.frameTimes.reduce((a, b) => a + b, 0) / metrics.frameTimes.length : 0;
        
        const avgFPS = avgFrameTime ? 1000 / avgFrameTime : 0;
        
        const slowFrames = metrics.frameTimes.filter(time => time > 16.67).length;
        const verySlowFrames = metrics.frameTimes.filter(time => time > 50).length;
        
        // Display results
        console.log('===== PERFORMANCE TEST RESULTS =====');
        console.log(`Test type: ${metrics.scrollPattern} scroll`);
        console.log(`Duration: ${(performance.now() - metrics.startTime) / 1000} seconds`);
        console.log(`Total scroll events: ${metrics.scrollEvents}`);
        console.log(`Average scroll handler time: ${avgScrollTime.toFixed(2)}ms`);
        console.log(`Frame count: ${metrics.frames}`);
        console.log(`Average FPS: ${avgFPS.toFixed(1)}`);
        console.log(`Slow frames (>16.67ms): ${slowFrames} (${((slowFrames / metrics.frameTimes.length) * 100).toFixed(1)}%)`);
        console.log(`Very slow frames (>50ms): ${verySlowFrames} (${((verySlowFrames / metrics.frameTimes.length) * 100).toFixed(1)}%)`);
        console.log('====================================');
        
        // Remove indicator
        const indicator = document.getElementById('performance-test-indicator');
        if (indicator) indicator.remove();
        
        return {
            testType: metrics.scrollPattern,
            duration: (performance.now() - metrics.startTime) / 1000,
            scrollEvents: metrics.scrollEvents,
            avgScrollTime: avgScrollTime.toFixed(2),
            frameCount: metrics.frames,
            avgFPS: avgFPS.toFixed(1),
            slowFrames,
            verySlowFrames,
            slowFramesPercentage: ((slowFrames / metrics.frameTimes.length) * 100).toFixed(1) + '%',
            verySlowFramesPercentage: ((verySlowFrames / metrics.frameTimes.length) * 100).toFixed(1) + '%'
        };
    }
    
    // Register functions globally
    window.startFloatingBlocksTest = startTest;
    
    console.log('FloatingBlocks performance testing initialized.');
    console.log('Run a test with:');
    console.log('  window.startFloatingBlocksTest("slow")     // Smooth scroll test');
    console.log('  window.startFloatingBlocksTest("fast")     // Fast oscillating scroll test');
    console.log('  window.startFloatingBlocksTest("erratic")  // Random jumps test');
    console.log('You can specify duration in ms as a second parameter, e.g.:');
    console.log('  window.startFloatingBlocksTest("slow", 5000)  // 5 second test');
})();
```

</p>
</details> 


--- 


> The "opacity" at the start: 

The opacity is not supposed to change **after** initialization (it could have been handled by applying class, like `o_isInitialised` or whatever). I moved it in start because, again,  I wrongly assumed that having it in `dynamicContent` would have impacted performance somehow, but apparently It's not the case.

> Actually added a comment on that opacity: do we really need it?

I think that we need it, yes. The snippet is designed to fulfill the entire container and so it's very visible.
Without this on page load  the scale transformation gives the feeling that the website is flickering.

I thought to use a CSS transition to be performed on the first scale operation only, just after initialization. It's a bit tricky because the transition should be removed *after* that it has been performed the very first time... otherwise it will impact the scrolling effect. Maybe `.once` event ? Not sure.

> Your last commit contained CSS to change the grid-area in edit mode... I did not understand it: with it it is broken, without it, it seems to act fine?

It's not "broken",  the problem is on the UX side because, in this context, you have the feeling that grid-items cannot move below the last visible grid line.  I think that's because we clearly see the limits due to the card, so we expect to see the entire drop-area while dragging... while normally the container's height can extend endlessly.
With this change, the end of the grid appears "cropped" at the bottom: I think that's the right behavior since the card size depends on the viewport... it better clarify the user what happens IMO.
Does it make sense for you? 
Normally I don't like these kind of "hacks", but at the end of the day this it's purely informative and harmless IMO.

**Without the change**

https://github.com/user-attachments/assets/f0600a46-6c93-4a52-ae8b-2be38eb285ed 

**With the change**

https://github.com/user-attachments/assets/7cb07c5e-5146-4e1a-b9fb-00274c566861

I've temporarily re-introduced the grid edition change + fixed a little typo. 

Let me know, thanks!

qsm-odoo commented:
@stefanorigano 

>  I've test both implementation performances and your version is not just simpler, but also slightly faster indeed, thanks.

I actually thought my implementation would have been a bit slower but I am not surprised it is faster either: the point is that it's not that much important and it would have been negligible either way. Unless we really do complex things/animations, the smaller code the better.

> I still need to get used to this new system I think.
> [...]
> The opacity is not supposed to change after initialization 

It's now meant to be mostly a description of the *state* rather than a description of the *flow* (just like OWL). The performance consideration is mostly left up to the framework, and if we ever have a performance issue, there is a chance it is to be solved at framework level.

Instead of "At start, I have to set the opacity to 1", you now say "While this interaction is alive, the opacity is supposed to be 1". The moment and the way it happens is irrelevant. With your start implementation, it worked, but it was a mistake: the opacity was not removed on destroy (you removed it yourself in edit mode in the option, which you don't have to).

> need opacity

The opacity is fine, especially as a one-line code, I just did not see a big difference or flicker effect without it.

> grid [...] It's not "broken", the problem is on the UX side

I cannot play your videos for some reasons but I am guessing the behavior you wanted. But it really is broken on my computer:
![image](https://github.com/user-attachments/assets/0072c799-1bdf-410b-a20b-d8489adc6165)
1. You seem to imply you want the grid to be "cut" at the bottom -> it does not seem to be on my side.
2. It is "broken" on the right (horizontally) -> see the grid background does not cover the full grid with your change

Maybe it's a Firefox issue? :shrug: Did not tested elsewhere.

stefanorigano commented:
> performance. 

 I registered a diminution of slow frames from  1% to 3% , depending by the test.
Intel/OSX, Firefox (no cache). The page contained  5 `s_floating_blocks` snippets, each one with a video and 2 parallax cards.

<details><summary>Details</summary>
<p>

```
===== BEFORE ======

===== PERFORMANCE TEST RESULTS ===== debugger eval code:155:17
Test type: fast scroll debugger eval code:156:17
Duration: 10.03 seconds debugger eval code:157:17
Total scroll events: 314 debugger eval code:158:17
Average scroll handler time: 0.04ms debugger eval code:159:17
Frame count: 314 debugger eval code:160:17
Average FPS: 31.2 debugger eval code:161:17
Slow frames (>16.67ms): 237 (75.7%) debugger eval code:162:17
Very slow frames (>50ms): 64 (20.4%) debugger eval code:163:17
==================================== debugger eval code:164:17

 ===== PERFORMANCE TEST RESULTS ===== debugger eval code:155:17
Test type: slow scroll debugger eval code:156:17
Duration: 10.008 seconds debugger eval code:157:17
Total scroll events: 341 debugger eval code:158:17
Average scroll handler time: 0.11ms debugger eval code:159:17
Frame count: 341 debugger eval code:160:17
Average FPS: 34.0 debugger eval code:161:17
Slow frames (>16.67ms): 293 (86.2%) debugger eval code:162:17
Very slow frames (>50ms): 32 (9.4%) debugger eval code:163:17
==================================== debugger eval code:164:17

===== PERFORMANCE TEST RESULTS ===== debugger eval code:155:17
Test type: erratic scroll debugger eval code:156:17
Duration: 10.003 seconds debugger eval code:157:17
Total scroll events: 126 debugger eval code:158:17
Average scroll handler time: 0.20ms debugger eval code:159:17
Frame count: 404 debugger eval code:160:17
Average FPS: 40.3 debugger eval code:161:17
Slow frames (>16.67ms): 258 (64.0%) debugger eval code:162:17
Very slow frames (>50ms): 44 (10.9%) debugger eval code:163:17
==================================== debugger eval code:164:17


​===== AFTER ======

===== PERFORMANCE TEST RESULTS ===== debugger eval code:155:17
Test type: fast scroll debugger eval code:156:17
Duration: 10.013 seconds debugger eval code:157:17
Total scroll events: 308 debugger eval code:158:17
Average scroll handler time: 0.07ms debugger eval code:159:17
Frame count: 308 debugger eval code:160:17
Average FPS: 30.7 debugger eval code:161:17
Slow frames (>16.67ms): 223 (72.6%) debugger eval code:162:17
Very slow frames (>50ms): 71 (23.1%) debugger eval code:163:17
==================================== debugger eval code:164:17

 ===== PERFORMANCE TEST RESULTS ===== debugger eval code:155:17
Test type: slow scroll debugger eval code:156:17
Duration: 10.009 seconds debugger eval code:157:17
Total scroll events: 325 debugger eval code:158:17
Average scroll handler time: 0.08ms debugger eval code:159:17
Frame count: 325 debugger eval code:160:17
Average FPS: 32.4 debugger eval code:161:17
Slow frames (>16.67ms): 283 (87.3%) debugger eval code:162:17
Very slow frames (>50ms): 26 (8.0%) debugger eval code:163:17
==================================== debugger eval code:164:17

===== PERFORMANCE TEST RESULTS ===== debugger eval code:155:17
Test type: erratic scroll debugger eval code:156:17
Duration: 10.071 seconds debugger eval code:157:17
Total scroll events: 129 debugger eval code:158:17
Average scroll handler time: 0.11ms debugger eval code:159:17
Frame count: 412 debugger eval code:160:17
Average FPS: 40.8 debugger eval code:161:17
Slow frames (>16.67ms): 262 (63.7%) debugger eval code:162:17
Very slow frames (>50ms): 34 (8.3%) debugger eval code:163:17
==================================== debugger eval code:164:17

```

</p>
</details> 


> "While this interaction is alive, the opacity is supposed to be 1". The moment and the way it happens is irrelevant.

Clear, thanks for the explanation ( @mano-odoo , interesting hint for your stuff 😉 )


> Grid Stuff -> Maybe it's a Firefox issue? 🤷 Did not tested elsewhere.

Nah, I'm on Firefox as well and I can reproduce your issue by just using a different screen, it's matter of viewport size indeed, uff. 
My hack enforce the number of rows, but to so that I have to enforce the number of columns as well. I could fix that but I'm sure that another issue would popup eventually. 
We have to get rid of the hack indeed, you're right. Pushed a new tmp on top 

Thanks

qsm-odoo reviewed (approved):
@robodoo rebase-ff r+

robodoo commented:
Merge method set to rebase and fast-forward.

robodoo commented:
@stefanorigano @qsm-odoo staging failed: ci/runbot on 979af8f5e45e86c393d7c5b9164178d4580a0d46 (view more at https://runbot.odoo.com/runbot/build/77959098)

stefanorigano commented:
Hello @qsm-odoo ,
about the [failing tests](https://runbot.odoo.com/runbot/build/77959098),  I see that the test fails because of the use of `s_` pattern classes,  but also because of the use basic utility ones (eg. `align-self-end`).

I'm not sure if these errors are unrelated to my changes OR if actions are required on my side (add some classes in the whitelist maybe ?). Thanks.

qsm-odoo commented:
@stefanorigano Before looking at the test adaptation (and improvement now that you made one), I wanted to look at your snippet again to see if nothing had to be changed/fixed first. I spoke with Soukéina about it too :+1: 

Here are the conclusions, which would simplify the snippet a bit, keeping it the same and fix the grid behavior (making it the same as everywhere else, except the cards' heights are fixed here):

1. Remove all `overflow-hidden` you used: as always, they would create more bugs than they solve (are they solving something here? Did not see it)

2. You have a `section.d-flex > .row.w-100.flex-grow-1.h-lg-100.align-self-end.align-self-lg-auto` card structure, I suggest changing it into a `section.d-flex.d-lg-block > .row.w-100.align-self-end`: the point is to enable flex only when you need it (if I am correct: only in mobile when you want to align the row at the bottom) so that the grid preview behave correctly in desktop. Indeed, your row will not be forced to full-height anymore (neither by flex nor the h-100 class) and make grid rows be added automatically while dragging grid-item, still being forced into the card area. (+ the `flex-grow-1` was already not required and `align-self-lg-auto` becomes useless since flex would be disabled on desktop).

3. Is changing the gutter with g-X classes necessary? Can't you leave the default and/or play with grid paddings/gap which can be controlled with the editor?

4. Maybe tweaking the structure a bit more, from `section.d-flex.d-lg-block > .row.w-100.align-self-end` to `section.d-flex.d-lg-block > .container-fluid.align-self-end > .row`. That one I am not entirely sure: the extra div level might be good to avoid, but it's also good to not mess up with our grid system too much (removing the w-100 class is the idea). The idea is also that the row is compensating an horizontal padding, one you could set on the inner container instead of the card section, which is a bit more "conventional".

5. Finally and maybe it should have been the first point: maybe grid mode is not the best mode for this snippet at the moment (it might be with the future task of Soukéina?). At the moment: add a grid item "at the bottom of a card", save, make your screen-width smaller -> you grid item is gone. Indeed the card kept its aspect ratio, but not the inner grid (which may even have grown instead).

Tell me what you think :pray:

stefanorigano commented:
> Tell me what you think 🙏

@qsm-odoo , starting from the very end:

 >  Finally and maybe it should have been the first point: maybe grid mode is not the best mode for this snippet at the moment (it might be with the future task of Soukéina?). At the moment: add a grid item "at the bottom of a card", save, make your screen-width smaller -> you grid item is gone.
 
The snippet is supposed to be used with a Title and a CTA. You can add a short description if you wish, but that's it.
Cards must have all the same height and this height must be shorter than the viewport. I can't see any other way to make this effect work and, with these constrains,  we cannot possibly adapt the layout to whatever content in whatever **position**... regardless by flex or grid.

There are hundreds of ways to make a snippet fail, even in the most simple layout.
Huge font-size + small column => text overflows outside. We don't address these "issues" of course, because it's the predictable behavior and it's on the user to decide if what he see on screen match his need.
 
If we want this kind of stuff like WxxFLOW or SxxPIFY do, and we want them now, I don't see any other possible approach rather then let the user decide in these cases.

> Remove all overflow-hidden you used: as always, they would create more bugs than they solve (are they solving something here? Did not see it)

It's to clarify the user what's happening because, if we let the content overflow outside, you may think that's a bug.
To crop the content within the grid make it clear(er) the the card size is "fixed" and that you have to stay within the boundaries. It's not bullet-prof, but seems better than the alternative.
![image](https://github.com/user-attachments/assets/45f2694e-dcd1-4e6b-9bc8-fe131d1c7bfb)

> 2, 4

I'll check

qsm-odoo commented:
@stefanorigano 

> The snippet is supposed to be used with a Title and a CTA.

Is not that an argument in favor of not using grid mode? Grid mode might become "the better default mode" in "single-columns" layouts when Soukéina will do her responsive grid task but if the point is not to do anything complicated, why enabling grid mode by default?

> There are hundreds of ways to make a snippet fail, even in the most simple layout.

That's never an argument to allow a snippet to be even more easily "broken" :wink: In this case, given the use case you mentioned (title + CTA), it's way less likely to fail without grid mode -> so why use it? Same way it's not enabled in most snippets.

> We don't address these "issues" of course, because it's the predictable behavior

That's exactly the point here too: that would not be predictable behavior in this case. People already don't understand our grid mode in its normal state: here we add the fact that the grid container has a viewport-specific fixed height increasing the chances for grid mode to fail.

Anyways, if you prefer to keep the grid and rely on the fact it will become responsive and better suited for this use case following the future task of Soukéina, that's fine by me. It's not that important at the moment, the editor UI is too complex either way.

> To crop the content within the grid make it clear(er) the the card size is "fixed" and that you have to stay within the boundaries.

Maybe, but I think it would be clear enough with it overflowing and seeing you cannot resize the card too :shrug: It does not feel that important. However, `overflow: auto` will prevent dropdowns to overflow the card, and that's very important because that's definitely reported as a bug, always (e.g. the searchbar snippet).

> 2, 4: I'll check

:+1:

stefanorigano commented:
@qsm-odoo , I've pushed 2 tmp commits on top.

One applying the suggested changes, the other essentially enforcing the row number to the minimum LG can handle with our desired design. This will prevent the user to drop stuff in XXL and then get it lost in LG.

There drawbacks are of course, but all considered looks like a good compromise to me.
Thanks

stefanorigano commented:
@qsm-odoo , pushed a commit on top to adapt design and options according to the [previous changes](https://github.com/odoo/odoo/pull/189299#issuecomment-2798500980).  In detail:
- adapt editor options according to the new structure (`data-exclude`, selectors,  etc.)
- introduced CSS `overflow: hidden;` rule in edit mode only. This will prevent the user to move grid-items outside the desired grid size.  No overflow is applied once the page is saved, allowing eventual "dynamic" content (eg. dropdowns) to overflow.
- renamed CSS variables to follow our general nomenclature (eg. `-desktop` => `-lg`)
- padding and aspect-ratio adjustments
- removed `overflow-hidden` classes (it was already the case, here for clarity)
- removed `gap-` utility classes (it was already the case, here for clarity)

---
> The snippet is supposed to be used with a Title and a CTA.
    > Is not that an argument in favor of not using grid mode?

I don't see how 🤔
You're supposed to use a limited about of elements here, yes, but you still want to move them around.
In our first tests we were using flex, no grid, and to move these elements by hitting [space] or [enter] wasn't efficient and prone to errors (beside being very frustrating).
Also it doesn't solve the issue of the overflowing content when the page resize 🤔

> Anyways, if you prefer to keep the grid and rely on the fact it will become responsive and better suited for this use case following the future task of Soukéina, that's fine by me.

Yes, thanks, it's the way to go for me.

qsm-odoo commented on addons/website/static/src/snippets/s_floating_blocks/options.xml:
```diff
@@ -10,14 +10,16 @@
 
-            <div class="o_grid_item g-height-2 g-col-lg-3 col-lg-3" style="z-index: 2; --grid-item-padding-y: 0px; --grid-item-padding-x: 0px; grid-area: 5 / 1 / 7 / 4;">
-                <a href="#" title="" role="button" class="btn btn-lg btn-secondary">Button</a>
+                <div class="o_grid_item g-col-lg-4 col-lg-4" style="z-index: 2; --grid-item-padding-y: 0px; --grid-item-padding-x: 0px; grid-area: 7 / 1 / 9 / 5;">
```
Good catch from Soukéina: `g-height-2` missing here (our test won't catch it since it is in a static file)

qsm-odoo commented on addons/website/views/snippets/s_floating_blocks.xml:
```diff
@@ -9,60 +9,66 @@
                     <section
                         data-name="Card"
-                        class="s_floating_blocks_block s_col_no_resize position-sticky d-flex parallax overflow-hidden py-5 o_cc o_cc5"
+                        class="s_floating_blocks_block s_col_no_resize position-sticky parallax py-5 py-lg-0 d-flex o_cc o_cc5"
```
@stefanorigano I push-forced with a rebase + squash of the TMP commit + 1 or 2 fix of mine... then I tested and the grid mode is still broken, pretty much like before. The solution I suggested fixed that, you implemented it slightly differently (e.g. not using the `d-lg-block` I suggested for instance), which is ok (since you use a container as suggested, why not centering it vertically with flex indeed)... but we need to find a solution that works :grimacing: I think the deviations you made that make it so the grid mode is not working are both "removing d-flex" and your "forced height on the row in css"... or at least the forced height but then the grid being random height + centered, the behavior is quite weird. See for instance, grid rows amount not following my mouse moves:
![image](https://github.com/user-attachments/assets/75b4c2bb-6c40-4438-bcd8-b50379869776)

Soukéina suggests that forcing the height on the container instead of the row might make it work, while keeping the flex center as you did. And adapting overflow hidden accordingly. Would you check? :pray:

stefanorigano replied:
Thanks @qsm-odoo  👍 

> See for instance, grid rows amount not following my mouse moves:
> Soukéina suggests that forcing the height on the container instead of the row might make it work[..]

Fixed, thanks @sobo-odoo , top 🙌💯 

> but then the grid being random height + centered, the behavior is quite weird.

It's the expected behavior of any grid shorter than its container no 🤔?
You can reproduce it on any other grid snippet using a viewport-relative min-height... maybe I didn't get it right?
I've pushed on top, let me know if there is something else, thanks! 
![image](https://github.com/user-attachments/assets/9e477382-18c4-4940-9c02-f229970d4892)

qsm-odoo replied:
>  You can reproduce it on any other grid snippet using a viewport-relative min-height... maybe I didn't get it right?

Never saw that before... that's awful :scream: The behavior when you are dragging and increasing the grid area is weird but quite ok... but when you are dragging and decreasing the size... that's awful, all items move when you release your mouse :|

Anyways, that's a problem for another time (cc @lebl-odoo)

qsm-odoo commented on addons/website/static/src/snippets/s_floating_blocks/floating_blocks.edit.js:
```diff
@@ -1,13 +0,0 @@
-    .add("website.floating_blocks", {
-        Interaction: FloatingBlocks,
-        mixin: FloatingBlocksEdit,
-    });
```
In fact, this was useless: I supposed there was a callToAction function you wanted to cancel but it does not exist at all

qsm-odoo reviewed (approved):
@robodoo rebase-ff r+
